Proxy Middleware: Add support for multiple backends, load balancing & healthchecks

nemothekid opened the pull request:
First off - thanks for creating caddy. We were looking for a solution that would allow us to write custom service discovery backends for our mesos cluster in nginx (like creating a custom upstream) as DNS-based service discovery wasn't working too well for us (failover & dns caching didn't work too well). It looked like we would have to write our module in C. With this we can write something simpler with Go.

As I have done work creating the reverse proxy backend (for our service discovery), I thought it would have made the most sense to provide more general features to the upstream backend (as we would use those too), that had most of the base features included, but still retained the ability to be easily extensible (any developer can simply copy the `staticUpstream` implementation and get load balancing & fail over, and be able to write & extend backends - for example doing manual DNS round robin.)

**Summary**
This PR builds off the `proxy` middleware to add the ability to have multiple backends, load balancing, (with `random`, `round_robin`, and `least_conn` directives), fail over, and health checks. I've attempted to emulate nginx features, but I haven't decided to completely emulate nginx outright.

**Backwards Compatibility**
These changes should still be backwards compatible, except for 1 thing. Whereas before if the config did not specify a scheme, caddy would match the backend scheme to the incoming request's scheme. Now, the scheme simply defaults to `http` if its not given.

**Config**
```
proxy from to [to [to...]] [{
    policy random | least_conn | round_robin
    fail_timeout duration
    max_fails integer
    health_check path duration
    proxy_header name value
}]
```

Multiple backends are specified with multiple `to` fields, and the block is entirely optional. The options in the block are as follows:

* `policy` - Load balancing policy. The default is `random`, which random selects a backend. `least_conn` selects a backend with the least amount of active connections, falling back to randomly selecting a backend if there are multiple such backends, and `round_robin` selects backends in a round robin fashion.
* `fail_timeout` - When a backend has failed, this timeout specifies how long to consider that backend as down. While it is down, requests will not be routed to that backend. A backend is "down" is caddy fails to communicate with it. The default value is `10s`. 
* `max_fails` - How many failures within `fail_timout` are needed to consider a backend as down. If `max_fails` is `0`, the failures will not cause the backend to be marked as down. The default value is `1`
* `health_check` - `path` and optional `duration`. Health check will check `path` on each backend, and if that backend returns a status code of 200-399, then that backend is healthy, if it doesn't the backend is marked as unhealthy for `duration` and no requests are routed to it. If this option is not provided then health checks are disabled. The default duration is `10s`
* `proxy_header`  - `name` and `value`. Sets headers to passed to backend. This option can be specified multiple times for multiple headers. You can also interpolate certain values as part of the headers with values provided to `Replacer` (also used by the `log` middleware)
  
**Examples**
Multiple backends
```
proxy / web1.local:80 web2.local:90 web3.local:100
```

Round Robin
```
proxy / web1.local:80 web2.local:90 web3.local:100 {
  policy round_robin
}
```

Health Checks & Proxy Headers
```
proxy / web1.local:80 web2.local:90 web3.local:100 {
  policy round_robin
  health_check /health
  proxy_header Host {host}
  proxy_header X-Caddy-When server-{when}
}
```

mholt commented:
This is amazing. You just turned Caddy into a load-balancing super hero - and with tests!

I am genuinely looking forward to trying this out. I love how easy it is to define multiple backends; and to  have multiple policies right off the bat is incredible. You also immediately resolved #39 which was opened at about the same time you submitted this.

I think your change is going to help a lot of people.

As for the backward compatibility, that is fine in this case; we're still pre-1.0 and the next release will be more than a patch.

Will look into this some more tomorrow!

mholt commented:
One quick question before I forget - if three remotes are specified, like your example `web1.local:80 web2.local:90 web3.local:100`, and a particular instance of Caddy happens to be serving web2.local:90, does that instance recognize that and proxy to the other 2 endpoints only?

I ask, because if each instance of Caddy could use the same Caddyfile, that would make deployment even easier. (Of course, the address of the server would be different in each Caddyfile, but the `import` directive can be used to share the rest of the file between them.)

mholt commented on middleware/proxy/proxy.go:
```diff
@@ -2,83 +2,181 @@
-	rules, err := parse(c)
-	if err != nil {
-		return nil, err
+func templateWithDelimiters(s string, vars requestVars) string {
```
I don't know if you had seen this, but the [replacer type](https://github.com/mholt/caddy/blob/master/middleware/replacer.go) in the middleware package does something similar. It's not a show-stopper for merging this PR but it's something to consider (even if it means enhancing the replacer type a little bit so it works with ).

nemothekid replied:
I hadn't seen it and I honestly wasn't sure how to best continue with string interpolation here which is why I went the route I did (I originally tried to use `text/template`). Using the already supplied `Replacer` is much cleaner IMO.

mholt commented on middleware/proxy/proxy.go:
```diff
@@ -2,83 +2,181 @@
+var tRe = regexp.MustCompile(`([\-\[\]()*\s])`)
+var tRe2 = regexp.MustCompile(`\$`)
+var openDelim = tRe2.ReplaceAllString(tRe.ReplaceAllString("{{", "\\$1"), "\\$")
+var closDelim = tRe2.ReplaceAllString(tRe.ReplaceAllString("}}", "\\$1"), "\\$")
```
For a consistent experience, I wonder if we could use a single { and } for template stuff, like how the [log package does it](https://github.com/mholt/caddy/blob/4ed93878019f2827956728f8f2c56f121de37f76/middleware/log/log.go#L121) - just two less keys to type. (The replacer type mentioned in my other comment uses {this format})

nemothekid commented:
Okay I've updated the branch per your suggestions.

1. I've removed the `{{upstream}}` and `{{upstream_host}}` variables. `{{upstream_host}}` is already technically provided to `r.Host`. If this variable is needed in the future it can be added in a new PR, I really wanted to keep the scope of this PR within the `proxy` middleware. However wrt to changes I made to `replacer`, I've opted to keep it private and simply expose the interface. In order to keep this feature `Replacer` needs a way for middlewares to extend the variables in `Replacer`.
2. In order to use `Replacer` in this middleware, I've changed the functionality of middleware a bit.
  * The `{remote}` replacer now checks the `X-Forwarded-For` header if its provided and uses that rather than `r.RemoteAddr`
  * The `responseRecorder` provided to `NewReplacer` can be `nil`, so I've checked that (proxy middleware didn't have a `responseRecorder`, and it didn't make much sense to me to allocate one).
  * There is a new `Replacer` interface that requires a `Replace(string) string` function. The `replacer` type wasn't exported and I opted to keep it that way.

mholt commented:
Looking great. The Replacer interface is a good idea. I was wondering when I made the type if it would be used with more than just logging someday.

mholt commented:
This package may need some work in the future. I will grant you push rights so it will be easier for you to help maintain it. You can still open a pull request, though, if you ever want a change reviewed or warn of a potential breaking change.

I'll merge this in now. Thank you again!

avelino commented:
+1

mholt commented:
Just curious, is it standard practice to default to http in reverse proxying if the scheme is not specified in the configuration? Or is it just a matter of preference? (I'm fine either way, I just didn't consider this way before.)

nemothekid commented:
Going off what [nginx does](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass) (the "standard practice") - if a protocol is not explicitly defined, a configuration error is thrown - so its not really a matter of defaults. I would have opted for that as well but, I'd rather not break backwards compatibility.

Now nginx is powerful enough that you can pass through the protocol (through *$variable_name*), but I'm not sure how useful that feature is. If your backend supports SSL, you will most likely want to use it for every request and not just select ones, and secondly, in most cases, the backend doesn't support SSL (like a node/rails app, that usually depends on the proxy server to implement ssl).

abiosoft commented:
Thanks @nemothekid

mholt commented:
Makes sense. I guess we can do that for now and if anyone complains, we can talk about it then.

j-mcnally commented:
Woah so now caddy is an haprxy replacement too? All of the Internet points!

j-mcnally commented:
A lot of ssl is terminated at the load balancer / reverse proxy. If it's on the same box that's probably fine, I'd prefer keeping ssl end to end, but a lot of servers aren't setup to handle it making http a good default. However maybe don't have a default and just require a protocol. It's not hard to specify one.

mholt commented:
@j-mcnally Been thinking a lot about this. Will probably let it slide for now as HTTP but if people start to speak up about it, we can reconsider (the joys of being pre-1.0)
